RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc

```diff
@@ -0,0 +1,286 @@
+/** \file ME0SegmentMatcher.cc
+ *
+ * \author David Nash
+ */
+
+#include <RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.h>
+
+#include <FWCore/PluginManager/interface/ModuleDef.h>
+#include <FWCore/Framework/interface/MakerMacros.h>
+
+#include <DataFormats/Common/interface/Handle.h>
+#include <FWCore/MessageLogger/interface/MessageLogger.h> 
+
+#include <DataFormats/MuonReco/interface/ME0Muon.h>
+
+#include "DataFormats/GeometryVector/interface/GlobalPoint.h"
+
+#include "TrackingTools/Records/interface/TrackingComponentsRecord.h"
+
+#include "DataFormats/GeometrySurface/interface/Plane.h"
+#include "TrackPropagation/SteppingHelixPropagator/interface/SteppingHelixPropagator.h"
+#include "TrackPropagation/SteppingHelixPropagator/interface/SteppingHelixStateInfo.h"
+#include "DataFormats/Math/interface/deltaR.h"
+
+
+#include "DataFormats/GeometrySurface/interface/LocalError.h"
+
+
+#include "TLorentzVector.h"
+
+#include "DataFormats/TrajectoryState/interface/LocalTrajectoryParameters.h"
+#include "TrackingTools/AnalyticalJacobians/interface/JacobianCartesianToLocal.h"
+#include "TrackingTools/AnalyticalJacobians/interface/JacobianLocalToCartesian.h"
+
+
+#include "Geometry/GEMGeometry/interface/ME0Geometry.h"
+#include <Geometry/Records/interface/MuonGeometryRecord.h>
+#include "Geometry/GEMGeometry/interface/ME0EtaPartitionSpecs.h"
+#include "Geometry/CommonTopologies/interface/StripTopology.h"
+#include <DataFormats/GeometrySurface/interface/SimpleDiskBounds.h>
+
+ME0SegmentMatcher::ME0SegmentMatcher(const edm::ParameterSet& pas) {
+  produces<std::vector<reco::ME0Muon> >();  
+  theX_PULL_CUT   = pas.getParameter<double>("maxPullX");
+  theX_RESIDUAL_CUT   = pas.getParameter<double>("maxDiffX");
+  theY_PULL_CUT   = pas.getParameter<double>("maxPullY");
+  theY_RESIDUAL_CUT   = pas.getParameter<double>("maxDiffY");
+  thePHIDIR_RESIDUAL_CUT   = pas.getParameter<double>("maxDiffPhiDirection");
+  //Might need to replace "OurSegments" with an edm::InputTag of "OurSegments"
+  OurSegmentsTag = pas.getParameter<edm::InputTag>("me0SegmentTag");
+  generalTracksTag = pas.getParameter<edm::InputTag>("tracksTag");
+  OurSegmentsToken_ = consumes<ME0SegmentCollection>(OurSegmentsTag);
+  generalTracksToken_ = consumes<reco::TrackCollection>(generalTracksTag);
+}
+
+ME0SegmentMatcher::~ME0SegmentMatcher() {}
+
+void ME0SegmentMatcher::produce(edm::Event& ev, const edm::EventSetup& setup) {
+
+
+    //Getting the objects we'll need
+    using namespace edm;
+
+    ESHandle<MagneticField> bField;
+    setup.get<IdealMagneticFieldRecord>().get(bField);
+    //ESHandle<Propagator> shProp;
+    //setup.get<TrackingComponentsRecord>().get("SteppingHelixPropagatorAlong", shProp);
+    const SteppingHelixPropagator* ThisshProp;
+    ThisshProp = new SteppingHelixPropagator(&*bField,alongMomentum);
```

Comment:
take the propagator from the EventSetup. The default-constructed one is likely not exactly what you want.
Also, ThisshProp is never deleted, creating a noticeable leak.